src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,168 @@
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
+
+  /*
+  * Retrieve the list of AstNodes from the AST body.
+  * Return empty seq if body is null or is not a list.
+  */
+    private def getChildrenList(ast: Ast): Seq[AstNode] = {
+      val body: AstNode = ast.getAttribute("body")
+      if(body == null) Seq.empty
+      else {
+        body match {
+          case list: AstList => list.asScala.toVector
+          case _ => Seq.empty
+        }
+      }
+    }
+
+    /**
+     * Generate recursively all descendant Scopes for the workflow.
+     * Should only be called during apply with a root Ast
+     * @param node current working node
+     * @param namespaces namespaces
+     * @param tasks tasks
+     * @param wdlSyntaxErrorFormatter formatter
+     * @return Some(instance of Scope) if the node has a Scope equivalent, None otherwise
+     */
+    def generateScopeTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Option[Scope] = {
+      node match {
+          //WORKFLOW
+        case w: Ast if w.getName.equals(AstNodeName.Workflow) =>
+          val workflowChildren = getChildrenList(w) flatMap (generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter))
+          val workflow: Workflow = Workflow(w, wdlSyntaxErrorFormatter)
+          workflow.setChildren(workflowChildren)
+          Some(workflow)
+
+          //CALL
+        case c: Ast if c.getName.equals(AstNodeName.Call) => Some(Call(c, namespaces, tasks, wdlSyntaxErrorFormatter))
+
+          //SCATTER
+        case s: Ast if s.getName.equals(AstNodeName.Scatter) =>
+          // Need to create the Scatter before generating its children to have the correct index layout order
+          val scatter = Scatter(s, scatterCount)
+          scatterCount += 1
+          val scatterChildren = getChildrenList(s) flatMap (generateScopeTree(_, namespaces, tasks, wdlSyntaxErrorFormatter))
+          scatter.setChildren(scatterChildren)
+          Some(scatter)
+
+        /* Cases need to be added here to support new scopes */
+        case _ => None
+      }
+    }
+  }
+
+  /**
+   * Generate a Scope Tree from an Ast Node.
+   * Node must match a Scope (currently: Workflow, Scatter or Scope)
+   * @param node node to build tree from
+   * @param namespaces namespaces used
+   * @param tasks tasks
+   * @param wdlSyntaxErrorFormatter wdl formatter
+   * @throws UnsupportedOperationException if node does not match any supported Scopes
+   * @return generated Scope Tree.
+   */
+  def generateTree(node: AstNode, namespaces: Seq[WdlNamespace], tasks: Seq[Task], wdlSyntaxErrorFormatter: WdlSyntaxErrorFormatter): Scope = {
+    new ScopeTreeGenerator().generateScopeTree(node, namespaces, tasks, wdlSyntaxErrorFormatter)
+    .getOrElse(throw new UnsupportedOperationException("Input node cannot be instantiated as a Scope"))
```

Comment:
The `ScopeTreeGenerator` class and its usage feels kind of awkward, it seems it exists only because of `scatterCount`?  If all that stuff were moved to this `Scope` object and the equivalent of the `generateScopeTree` method were parameterized by and returned a scatterCount, I think you could do away with `ScopeTreeGenerator` and that `var`.